### PR TITLE
Add print_annotated_stack helper function 

### PR DESCRIPTION
**Pull request by nhatsmrt**
This helper function shows the content of the stack trace, optionally annotating each entry with the appropriate tag. Currently, it only annotates symbols and slab cache objects.

Example output:

```
#1 [ffffbbdfc1497ec0] context_switch at ffffffffa304efe3
#2 [ffffbbdfc1497ec0] __schedule at ffffffffa304efe3
ffffbbdfc1497ec0: 00007f581782eca8
ffffbbdfc1497ec8: function symbol: handle_mm_fault+0xba
ffffbbdfc1497ed0: 0000000000000004
ffffbbdfc1497ed8: 73465a68e2eef000
ffffbbdfc1497ee0: slab object: task_struct
ffffbbdfc1497ee8: ffffbbdfc1497f58
ffffbbdfc1497ef0: 0000000000000000
ffffbbdfc1497ef8: 0000000000000000
ffffbbdfc1497f00: 0000000000000000
ffffbbdfc1497f08: slab object: task_struct
ffffbbdfc1497f10: function symbol: schedule+0x94
#3 [ffffbbdfc1497f18] schedule at ffffffffa304f0d8
ffffbbdfc1497f18: slab object: task_struct
ffffbbdfc1497f20: ffffbbdfc1497f48
ffffbbdfc1497f28: function symbol: __x64_sys_pause+0x26
#4 [ffffbbdfc1497f30] __do_sys_pause at ffffffffa2c61d65
ffffbbdfc1497f30: 0000000000000000
ffffbbdfc1497f38: function symbol: do_syscall_64+0x70
#5 [ffffbbdfc1497f40] do_syscall_x64 at ffffffffa304a7d3
#6 [ffffbbdfc1497f40] do_syscall_64 at ffffffffa304a7d3
ffffbbdfc1497f40: 0000000000000000
ffffbbdfc1497f48: 0000000000000000
ffffbbdfc1497f50: symbol: entry_SYSCALL_64+0x99
#7 [ffffbbdfc1497f58] entry_SYSCALL_64 at ffffffffa3200099
ffffbbdfc1497f58: 00007f58172aed08
```

### CODE REVIEW

**Comment by pfactum** [comment]
Just for reference, this is how I implement this in [`crush`](https://codeberg.org/pf-kernel/crush): [backtrace.py#L58](https://codeberg.org/pf-kernel/crush/src/branch/master/crush/command/backtrace.py#L58).

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:31]: Interrupted frames essentially begin an entirely new stack. So rather than ignoring them, you should treat them almost like separators. So, e.g., if you have frame 0-2 with `interrupted = False`, then frame 3 with `interrupted = True`, then frame 4 with `interrupted = False`, you pretty much have to dump two stacks: one for frames 0-2, and one for frames 3-4.

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:51]: Maybe do something like `byteorder = "little" if (prog.platform.flags & PlatformFlags.IS_LITTLE_ENDIAN) else "big"` up where you get `word_size` instead of doing this on every iteration.

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:54]: FYI, these stack frames with the same address are typically because of inlined functions.

**Review by osandov — COMMENTED** [review]
Left a couple of comments about the implementation. In addition, I have some ideas about improving the output format:

1. A header explaining the output columns would be useful for anyone who isn't immediately familiar with the output.
2. I don't love crash's way of formatting the beginning of the stack frame. Maybe the `str()` representation of `drgn.StackFrame` would be better, although it might be too verbose.
3. For addresses which we're able to identify, let's include the address and the identity.

As a starting point, this:

```diff
diff --git a/drgn/helpers/common/stack.py b/drgn/helpers/common/stack.py
index f3f788bf..26c7e791 100644
--- a/drgn/helpers/common/stack.py
+++ b/drgn/helpers/common/stack.py
@@ -39,6 +39,7 @@ def print_annotated_stack(trace: StackTrace) -> None:
     stack_bytes = prog.read(start_addr, stack_size)
     frame_ind = 0
 
+    print("STACK POINTER     VALUE")
     for offset in range(0, len(stack_bytes), word_size):
         addr = start_addr + offset
         hex_addr = hex_format.format(addr)
@@ -54,23 +55,14 @@ def print_annotated_stack(trace: StackTrace) -> None:
         # for e.g, the first context_switch frame
         while frame_ind < len(frames_addrs) and addr == frames_addrs[frame_ind]:
             frame = frames[frame_ind]
-
-            name = frame.name
-            if name is None:
-                try:
-                    name = frame.symbol().name
-                except LookupError:
-                    name = hex(frame.pc)
-
-            instruction_addr = hex_format.format(frame.register("rip"))
-
-            print(f"#{frame_ind + 1} [{hex_addr}] {name} at {instruction_addr}")
+            print(f"[stack frame {frame}]")
 
             frame_ind += 1
         else:
-            formatted_value = identify_address(prog, word_val)
+            formatted_value = hex_format.format(word_val)
 
-            if formatted_value is None:
-                formatted_value = hex_format.format(word_val)
+            identified = identify_address(prog, word_val)
+            if identified:
+                formatted_value += f" [{identified}]"
 
             print(f"{hex_addr}: {formatted_value}")
```

Gives us:

```
STACK POINTER     VALUE
[stack frame #0 at 0xffffffffa0e77d6e (__schedule+0x436/0x495) in context_switch at ./kernel/sched/core.c:5182:2 (inlined)]
[stack frame #1 at 0xffffffffa0e77d6e (__schedule+0x436/0x495) in __schedule at ./kernel/sched/core.c:6494:8]
ffff9a7800013d28: fffff170000d0002
ffff9a7800013d30: 00000000a0a486f8
ffff9a7800013d38: ffffffffa0a48dfe [function symbol: flush_tlb_func+0xee]
ffff9a7800013d40: 0000000000000004
ffff9a7800013d48: c72e0e8e1cad2100
ffff9a7800013d50: ffff924441178000 [slab object: task_struct]
ffff9a7800013d58: ffff924441178000 [slab object: task_struct]
ffff9a7800013d60: ffff9a7800013e10
ffff9a7800013d68: ffff924441177ff0 [slab object: mm_struct]
ffff9a7800013d70: ffff924441178000 [slab object: task_struct]
ffff9a7800013d78: ffff924441178000 [slab object: task_struct]
ffff9a7800013d80: ffffffffa0e77e56 [function symbol: schedule+0x89]
[stack frame #2 at 0xffffffffa0e77e56 (schedule+0x89/0xc7) in schedule at ./kernel/sched/core.c:6570:3]
ffff9a7800013d88: ffff9a7800013de8
ffff9a7800013d90: 0000000000000000
ffff9a7800013d98: ffffffffa0a5803f [function symbol: do_wait+0x231]
[stack frame #3 at 0xffffffffa0a5803f (do_wait+0x231/0x2e3) in do_wait at ./kernel/exit.c:1537:4]
ffff9a7800013da0: ffff924441178478 [slab object: task_struct]
ffff9a7800013da8: ffff924441178000 [slab object: task_struct]
ffff9a7800013db0: 0000000000000004
ffff9a7800013db8: 0000000000000000
ffff9a7800013dc0: 00007ffc8272d0a0
ffff9a7800013dc8: 0000000000000000
ffff9a7800013dd0: fffffffffffffffd
ffff9a7800013dd8: 0000000000000004
ffff9a7800013de0: ffffffffa0a58e93 [function symbol: kernel_wait4+0xc2]
[stack frame #4 at 0xffffffffa0a58e93 (kernel_wait4+0xc2/0x11b) in kernel_wait4 at ./kernel/exit.c:1681:8]
ffff9a7800013de8: 0000000400000004
ffff9a7800013df0: 0000000000000000
ffff9a7800013df8: 0000000000000000
ffff9a7800013e00: ffff9a7800000000
ffff9a7800013e08: 0000000000000000
ffff9a7800013e10: 0000000000000000
ffff9a7800013e18: ffff924441178000 [slab object: task_struct]
ffff9a7800013e20: ffffffffa0a5735f [function symbol: child_wait_callback+0x0]
ffff9a7800013e28: ffff924441188028 [slab object: signal_cache]
ffff9a7800013e30: ffff924441188028 [slab object: signal_cache]
ffff9a7800013e38: 0000563900000000
ffff9a7800013e40: c72e0e8e1cad2100
ffff9a7800013e48: 0000000000000000
ffff9a7800013e50: ffff9a7800013f30
ffff9a7800013e58: ffff9a7800013f58
ffff9a7800013e60: 0000000000000000
ffff9a7800013e68: 0000000000000000
ffff9a7800013e70: 0000000000000000
ffff9a7800013e78: ffffffffa0a58f24 [function symbol: __do_sys_wait4+0x38]
[stack frame #5 at 0xffffffffa0a58f24 (__do_sys_wait4+0x38/0x8c) in __do_sys_wait4 at ./kernel/exit.c:1709:13]
ffff9a7800013e80: ffffffffa0e7bd39 [function symbol: _raw_spin_unlock_irq+0x10]
ffff9a7800013e88: ffffffffa0a630aa [function symbol: do_sigaction+0xf8]
ffff9a7800013e90: ffff924441180020 [slab object: sighand_cache]
ffff9a7800013e98: 0000000000000255
ffff9a7800013ea0: c72e0e8e1cad2100
ffff9a7800013ea8: 0000000000000002
ffff9a7800013eb0: 00007ffc8272cee0
ffff9a7800013eb8: ffff9a7800013f58
ffff9a7800013ec0: 0000000000000000
ffff9a7800013ec8: 0000000000000000
ffff9a7800013ed0: 0000000000000046
ffff9a7800013ed8: ffff924441178000 [slab object: task_struct]
ffff9a7800013ee0: ffff924441178000 [slab object: task_struct]
ffff9a7800013ee8: ffff9a7800013f58
ffff9a7800013ef0: 0000000000000000
ffff9a7800013ef8: ffffffffa0a261f0 [function symbol: fpregs_assert_state_consistent+0x1b]
ffff9a7800013f00: 0000000000000000
ffff9a7800013f08: ffffffffa0ab0b3f [function symbol: exit_to_user_mode_prepare+0xa6]
ffff9a7800013f10: 0000000000000000
ffff9a7800013f18: c72e0e8e1cad2100
ffff9a7800013f20: ffff9a7800013f58
ffff9a7800013f28: 0000000000000000
ffff9a7800013f30: ffff9a7800013f48
ffff9a7800013f38: ffffffffa0e72a43 [function symbol: do_syscall_64+0x70]
[stack frame #6 at 0xffffffffa0e72a43 (do_syscall_64+0x70/0x8a) in do_syscall_x64 at ./arch/x86/entry/common.c:50:14 (inlined)]
[stack frame #7 at 0xffffffffa0e72a43 (do_syscall_64+0x70/0x8a) in do_syscall_64 at ./arch/x86/entry/common.c:80:7]
ffff9a7800013f40: 0000000000000000
ffff9a7800013f48: 0000000000000000
ffff9a7800013f50: ffffffffa100009b [symbol: entry_SYSCALL_64+0x9b]
[stack frame #8 at 0xffffffffa100009b (entry_SYSCALL_64+0x9b/0x197) at ./arch/x86/entry/entry_64.S:120]
ffff9a7800013f58: 0000000000000000
```

What do you think? I'd especially love suggestions on how to format the beginnings of stack frames more nicely.

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py:54]: Makes sense!

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py:51]:
```
drgn/helpers/common/stack.py:54: error: Argument 2 to "from_bytes" of "int" has incompatible type "str"; expected "Union[Literal['little'], Literal['big']]"
```

Something like this? Also `Literal` is Python 3.8 > so if I try import the type for annotating:

```
from typing import Literal, Union
```

It would also scream at me:
```
drgn/helpers/common/stack.py:12: error: Module "typing" has no attribute "Literal"; maybe "_Literal"?
```

**Comment by nhatsmrt** [comment]
New annotated stack trace:
```
STACK POINTER     VALUE
[stack frame #0 at 0xffffffff9c477fce (__schedule+0x436/0x495) in context_switch at ./kernel/sched/core.c:5182:2 (inlined)]
[stack frame #1 at 0xffffffff9c477fce (__schedule+0x436/0x495) in __schedule at ./kernel/sched/core.c:6494:8]
ffffb58ec133bdc0: ffffa39d02e1f458 [slab object: radix_tree_node]
ffffb58ec133bdc8: 0000000000000000
ffffb58ec133bdd0: 0000000000000000
ffffb58ec133bdd8: 0000000000000004
ffffb58ec133bde0: 926299c3d4a71600
ffffb58ec133bde8: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133bdf0: 00007ffd0a853220
ffffb58ec133bdf8: 0000000000000000
ffffb58ec133be00: 0000000000000000
ffffb58ec133be08: 0000000000000000
ffffb58ec133be10: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133be18: ffffffff9c4780b6 [function symbol: schedule+0x89]
[stack frame #2 at 0xffffffff9c4780b6 (schedule+0x89/0xc7) in schedule at ./kernel/sched/core.c:6570:3]
ffffb58ec133be20: 0000000000000000
ffffb58ec133be28: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133be30: ffffffff9c47b309 [function symbol: schedule_hrtimeout_range_clock+0x3f]
[stack frame #3 at 0xffffffff9c47b309 (schedule_hrtimeout_range_clock+0x3f/0xc8) in schedule_hrtimeout_range_clock at ./kernel/time/hrtimer.c:2296:3]
ffffb58ec133be38: 00007fa2e841e000
ffffb58ec133be40: 00007fa2e841e1c0
ffffb58ec133be48: 0000000000000b54
ffffb58ec133be50: ffffffff9c05f45a [function symbol: recalc_sigpending+0xe]
ffffb58ec133be58: ffffffff9c05fac9 [function symbol: dequeue_signal+0x16e]
ffffb58ec133be60: 0000000000000000
ffffb58ec133be68: 926299c3d4a71600
ffffb58ec133be70: 0000000000000000
ffffb58ec133be78: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133be80: 926299c3d4a71600
ffffb58ec133be88: 0000000000000000
ffffb58ec133be90: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133be98: 00007ffd0a853220
ffffb58ec133bea0: ffffffff9c062abb [function symbol: __x64_sys_rt_sigtimedwait+0x18f]
[stack frame #4 at 0xffffffff9c062abb (__x64_sys_rt_sigtimedwait+0x18f/0x23b) in freezable_schedule_hrtimeout_range at ./include/linux/freezer.h:250:13 (inlined)]
[stack frame #5 at 0xffffffff9c062abb (__x64_sys_rt_sigtimedwait+0x18f/0x23b) in do_sigtimedwait at ./kernel/signal.c:3604:9 (inlined)]
[stack frame #6 at 0xffffffff9c062abb (__x64_sys_rt_sigtimedwait+0x18f/0x23b) in __do_sys_rt_sigtimedwait at ./kernel/signal.c:3648:8 (inlined)]
[stack frame #7 at 0xffffffff9c062abb (__x64_sys_rt_sigtimedwait+0x18f/0x23b) in __se_sys_rt_sigtimedwait at ./kernel/signal.c:3626:1 (inlined)]
[stack frame #8 at 0xffffffff9c062abb (__x64_sys_rt_sigtimedwait+0x18f/0x23b) in __x64_sys_rt_sigtimedwait at ./kernel/signal.c:3626:1]
ffffb58ec133bea8: 0000000100000014
ffffb58ec133beb0: 7fffffffffffffff
ffffb58ec133beb8: 00007fa2e841e1c0
ffffb58ec133bec0: ffffffff9c17bfa1 [function symbol: handle_mm_fault+0xba]
ffffb58ec133bec8: 0000000000000000
ffffb58ec133bed0: ffffffffffffffff
ffffb58ec133bed8: 0000000000000014
ffffb58ec133bee0: 00007fa2e841e1c0
ffffb58ec133bee8: 0000000000000100
ffffb58ec133bef0: 0000000000000046
ffffb58ec133bef8: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133bf00: ffffa39d03588000 [slab object: task_struct]
ffffb58ec133bf08: 926299c3d4a71600
ffffb58ec133bf10: 0000000000000000
ffffb58ec133bf18: ffffb58ec133bf48
ffffb58ec133bf20: ffffb58ec133bf58
ffffb58ec133bf28: 0000000000000000
ffffb58ec133bf30: 0000000000000000
ffffb58ec133bf38: ffffffff9c472ca3 [function symbol: do_syscall_64+0x70]
[stack frame #9 at 0xffffffff9c472ca3 (do_syscall_64+0x70/0x8a) in do_syscall_x64 at ./arch/x86/entry/common.c:50:14 (inlined)]
[stack frame #10 at 0xffffffff9c472ca3 (do_syscall_64+0x70/0x8a) in do_syscall_64 at ./arch/x86/entry/common.c:80:7]
ffffb58ec133bf40: 0000000000000000
ffffb58ec133bf48: 0000000000000000
ffffb58ec133bf50: ffffffff9c60009b [symbol: entry_SYSCALL_64+0x9b]
[stack frame #11 at 0xffffffff9c60009b (entry_SYSCALL_64+0x9b/0x197) at ./arch/x86/entry/entry_64.S:120]
ffffb58ec133bf58: 00007fa2e967bd08
[stack frame #12 at 0x7fa2e84498dc]
```

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:100]: Make sure that this is still aligned nicely on 32-bit, since those stack pointers are only 8 characters wide. Maybe we can just pad the stack pointer with extra spaces on 32-bit.

**Review by osandov — CHANGES_REQUESTED** [review]

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Fixed!

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Fixed!

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: If it helps, one thing I could do is to put this in its own function (_get_stack_pointer(frame)) and/or add some comment reminding us to change it into a generic stack pointer getter later.

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Hmmm I added some logic to handle this:
```
    if is_64_bit:
        print("STACK POINTER     VALUE")
    else:  # 8 extra spaces
        print("STACK POINTER             VALUE")
```
We have to figure out how to test it on a 32-bit platform tho 🤔

Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Hardcoding a bunch of things to test the alignment:
Screen Shot 2022-11-07 at 2 00 17 PM

Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: (Basically just added 8 spaces)
```
                if is_64_bit:
                    print(f"{hex_addr}: {formatted_value}")
                else:
                    # 8 extra spaces to align the value
                    print(f"{hex_addr}:         {formatted_value}")
```

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:37]: I didn't love the extra spaces for 32-bit when I looked at it. What do you think of this:

```
STACK     VALUE
POINTER
01234567: 012345678
01234567: 012345678
```

I.e., wrap the "STACK POINTER" heading so that it fits within the natural width. Something like:

```suggestion
    if prog.platform.flags & PlatformFlags.IS_64_BIT:
        print("STACK POINTER     VALUE")
    else:
        print("STACK     VALUE\nPOINTER")
```

You should consolidate this with the other `IS_64_BIT` checks if you can.

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:98]: We can simplify this a lot if we define the format for the line based on 64-bit vs 32-bit at the beginning of the function:

```python3
    if prog.platform.flags & PlatformFlags.IS_64_BIT:
        line_format = "{:016x}: {:016x}{}"
    else:
        line_format = "{:08x}: {:08x}{}"
```

Then all you need here is:

```suggestion
                identified = identify_address(prog, word_val)
                if identified is None:
                    identified = ""
                else:
                    identified = f" [{identified}]"
                print(line_format.format(addr, word_val, identified))
```

(Then you can get rid of `hex_format` and `hex_addr` above, too.)

File: tests/linux_kernel/helpers/test_stack.py
Comment (osandov) [inline tests/linux_kernel/helpers/test_stack.py:23]: Use `self.assertIn` for all of these checks so that we get a nicer message if this fails:

```suggestion
            self.assertIn("slab object: task_struct", printed_trace)
```

File: drgn/helpers/common/stack.py
Comment (osandov) [inline drgn/helpers/common/stack.py:9]: Grammar nit:

```suggestion
The ``drgn.helpers.common.stack`` module provides helpers for working with
stack traces.
```

**Review by osandov — CHANGES_REQUESTED** [review]
This is almost there. I left a few more comments inline. Additionally, the commit message for the first commit is misleading; `__len__` and `__iter__` are already implemented, we're just missing type annotations for them. A more accurate commit message would be:

```
Add missing type annotations for StackTrace dunder methods

StackTrace supports __len__() and __iter__(), but they aren't annotated.
Add them.

Fixes: 80c9fb35ff96 ("Add type hint stubs and generate documentation from them")
```

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py:87]: Hmm now that I look at it - probably not. Might be remnants of an earlier attempt that somehow survive the copy-paste-rewrite process...

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py:37]: Looks good to me - I don't have a strong preference on this, so I'll defer to you 😅

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Fixed!

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Fixed!

File: tests/linux_kernel/helpers/test_stack.py
Comment (nhatsmrt) [reply tests/linux_kernel/helpers/test_stack.py]: Fixed!

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Oops :P Fixed!

File: drgn/helpers/common/stack.py
Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Oh neat! That's much cleaner!

Comment (nhatsmrt) [reply drgn/helpers/common/stack.py]: Fixed to using this.

**Review by osandov — APPROVED** [review]
This is awesome. I expanded on the documentation/comments a bit and made some nitpicky style changes. Once tests pass, I'll merge it.

**Pull request merged**
